keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,444 @@
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager.
+
+Existing logic:
+1. When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
+  1. [Proposed change] When a user is sure that a node is shutdown, it will apply a `quarantine` taint on the node.
+  1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.
+    1. new `quarantine` taint: no execute effect, meaning it will evict pods and start on different node.
+
+1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachement` to be deleted.
+
+1. This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+1. When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.
+
+1. [Proposed change] The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.
```

Comment:
This might not necessary considering some driver (controller side) cannot verify mountpoint is being used. 

by adding quarantine taint, we are certain that no mount points should exist and being used?

Reply:
This was from previous version of the KEP. I'll remove this.